### PR TITLE
enable and fix existing and add some more tests CLI tests.

### PR DESCRIPTION
**Pull request by riuvshin**
### What does this PR do?
fix existing CLI tests and add some more:
- cmd config test
- cmd stop, restart tests
- cmd info test
- cmd backup / restore
- cmd offline

logs:

```
Running tests in container from image eclipse/che-bats:nightly
Running functional bats tests for CLI prompts and usage
 ✓ test CLI prompt to provide volume for docker sock
 ✓ test CLI prompt to provide directory for user data
 ✓ test CLI 'usage' when running container without command

3 tests, 0 failures
Running functional bats tests for init and destroy commands
 ✓ test 'init' and 'destroy --quiet' with existing dir
 ✓ test 'init' and 'destroy --quiet' with non-existing dir
 ✓ test 'init' and 'destroy --quiet --cli' with existing dir
 ✓ test 'init' and 'destroy --quiet --cli' with non-existing dir

4 tests, 0 failures
Running functionals bats tests for start command
 ✓ test cli 'start' command with default settings
 ✓ test cli 'stop' command with default settings
 ✓ test cli 'restart' command with default settings
 ✓ test cli 'start' with custom port

4 tests, 0 failures
Running functionals bats tests for info command
 ✓ test CLI 'info' command

1 test, 0 failures
Running functionals bats tests for config command
 ✓ test CLI 'config' command

1 test, 0 failures
Running functionals bats tests for backup / restore commands
 ✓ test cli 'backup' command: backup fail if che is running
 ✓ test cli 'restore' command: restore fail if che is running
 ✓ test cli 'restore' command: restore fail if no backup found
 ✓ test cli 'backup / restore' commands

4 tests, 0 failures
Running functionals bats tests for offline command
 ✓ test cli 'offline' command: with default parameters
 ✓ test cli 'offline' command: include custom stack images

2 tests, 0 failures
```

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/5303

#### Changelog
fix existing CLI tests and add some more

### CODE REVIEW

**Comment by riuvshin** [comment]
@mlabuda it would be awesome if you also will take a look on this :)

there were a few issues in existing tests:
1) in bash you can't compare strings using `-eq`  like here https://github.com/eclipse/che/blob/5.11.2/dockerfiles/cli/tests/cmd_start_tests.bats#L52  `-eq` works only with integers.
2) in `test_base.sh` clean up of testrun dir was broken and fixed with this:
https://github.com/eclipse/che/pull/5304/files#diff-2cd4118d0466159af0a0663766059fccL17

(it tried to remove folder with path of host, but it should use path of which is actual for container where tests run)

File: dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  [[ "$(docker inspect --format='{{.State.Running}}' chetest)" == "true" ]]
+  ip_address=$(docker inspect -f {{.NetworkSettings.Networks.bridge.IPAddress}} chetest)
+  curl -fsS http://${ip_address}:8080  > /dev/null
+
+  #WHEN
+  docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data -e CHE_CONTAINER=chetest $CLI_IMAGE stop --skip:nightly --skip:pull
+
+  #THEN
+  #check that state is not set because CLI stop command also removes container.
+  [[ "$(docker inspect --format='{{.State.Running}}' chetest)" == "" ]]
```
Comment (mlabuda) [inline dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats]: Is this necessary? In next step you check that there is no such container, thus this should always pass.

File: dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data -e CHE_CONTAINER=chetest $CLI_IMAGE start --skip:nightly --skip:pull
+  [[ "$(docker inspect --format='{{.State.Running}}' chetest)" == "true" ]]
+  ip_address=$(docker inspect -f {{.NetworkSettings.Networks.bridge.IPAddress}} chetest)
+  curl -fsS http://${ip_address}:8080  > /dev/null
+
+  #WHEN
+  docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data -e CHE_CONTAINER=chetest $CLI_IMAGE restart --skip:nightly --skip:pull
+
+  #THEN
+  [[ "$(docker inspect --format='{{.State.Running}}' chetest)" == "true" ]]
```
Comment (mlabuda) [inline dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats]: To verify that a container is running after performing restart command, I would rather choose more bulletproof check. This way, if restart would do nothing, container would still be in running state and those checks would pass.

I suggest to use e.g. .State.Pid value for check actual value of process id before and after restart. But bcs. pid is not unique, maybe better would be to use .State.StartedAt.

**Review by mlabuda — CHANGES_REQUESTED** [review]
At the time of writing those tests I was newborn to bash scripting, my bad for those bugs. 

PR review done. There are just 2 comments, otherwise LGTM.

File: dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats (hunk 1 above)
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats]: agreed

File: dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats (hunk 2 above)
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats]: good catch! I will use che container ID each restart it is unique

File: dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats [hunk 3]
```diff
@@ -71,18 +68,18 @@ teardown() {
...
   ip_address=$(docker inspect -f {{.NetworkSettings.Networks.bridge.IPAddress}} chetest)
   curl -fsS http://${ip_address}:8080  > /dev/null
+  che_container_id=$(docker inspect --format="{{.Id}}" chetest)
 
   #WHEN
   docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data -e CHE_CONTAINER=chetest $CLI_IMAGE restart --skip:nightly --skip:pull
 
   #THEN
-  [[ "$(docker inspect --format='{{.State.Running}}' chetest)" == "true" ]]
+  [[ "$(docker inspect --format="{{.Id}}" chetest)" != "$che_container_id" ]]
```
Comment (mlabuda) [inline dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats]: Does the restart command on docker CLI container delete a container and start a new one? If not and it just reuse `docker restart` with some more magic, then the container ID is still the same.

**Review by mlabuda — CHANGES_REQUESTED** [review]

File: dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats (hunk 3 above)
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_start_stop_restart_tests.bats]: yes, CLI restart command does CLI stop and CLI start commands, and CLI stop command stop and remove container.

**Comment by riuvshin** [comment]
@mlabuda @benoitf @TylerJewell hey guys, I've updated PR with fixes and more tests. please review

File: dockerfiles/cli/tests/cmd_backup_restore_tests.bats [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  #GIVEN
+  if [ ! port_is_free 8080 ]; then
+    [ "$status" -eq 1 ]
+    [ "$output" = "Default port 8080 for che server is used. Cannot run this test on default che server port" ]
+  fi
+  tmp_path="${TESTRUN_DIR}"/cli_cmd_backup_fail_if_che_is_running
+  mkdir -p "${tmp_path}"
+  docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data $CLI_IMAGE start --skip:nightly --skip:pull
+  [[ "$(docker inspect --format='{{.State.Running}}' che)" == "true" ]]
+  ip_address=$(docker inspect -f {{.NetworkSettings.Networks.bridge.IPAddress}} che)
```
Comment (mlabuda) [inline dockerfiles/cli/tests/cmd_backup_restore_tests.bats]: we are using those 3 lines on many places. Maybe we could extract them and parametrize to make it reusable. WDYT?

File: dockerfiles/cli/tests/cmd_config_tests.bats [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#
+# Contributors:
+#   Roman Iuvshyn
+
+load '/bats-support/load.bash'
+load '/bats-assert/load.bash'
+source /dockerfiles/cli/tests/test_base.sh
+
+teardown() {
+  docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data $CLI_IMAGE destroy --quiet --skip:nightly --skip:pull
```
Comment (mlabuda) [inline dockerfiles/cli/tests/cmd_config_tests.bats]: See comment about setup/teardown above.

File: dockerfiles/cli/tests/cmd_config_tests.bats [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  [[ -e "${container_tmp_path}"/cli.log ]]
+  [[ -d "${container_tmp_path}"/instance ]]
+  [[ -d "${container_tmp_path}"/instance/config ]]
+  [[ -f "${container_tmp_path}"/instance/config/che.env ]]
+  [[ -d "${container_tmp_path}"/instance/data ]]
+  [[ -d "${container_tmp_path}"/instance/logs ]]
+  [[ -d "${container_tmp_path}"/instance/stacks ]]
+  [[ -d "${container_tmp_path}"/instance/templates ]]
+  [[ -f "${container_tmp_path}"/instance/docker-compose-container.yml ]]
+  [[ -f "${container_tmp_path}"/instance/che.ver.do_not_modify ]]
```
Comment (mlabuda) [inline dockerfiles/cli/tests/cmd_config_tests.bats:35]: just a question, why do you check some files with -f and others with -e ?

**Review by mlabuda — CHANGES_REQUESTED** [review]

File: dockerfiles/cli/tests/cmd_config_tests.bats (hunk 6 above)
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_config_tests.bats:35]: there is no `-e` checks, `-d` - test is it a dir and `-f` is it a file

File: dockerfiles/cli/tests/cmd_backup_restore_tests.bats (hunk 4 above)
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_backup_restore_tests.bats]: +1 will do

File: dockerfiles/cli/tests/cmd_backup_restore_tests.bats [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+source /dockerfiles/cli/tests/test_base.sh
+
+# Kill running che server instance if there is any to be able to run tests
+setup() {
+  kill_running_named_container che
+  remove_named_container che
+}
+
+teardown() {
+  docker run --rm -v "${SCRIPTS_DIR}":/scripts/base -v /var/run/docker.sock:/var/run/docker.sock -v "${tmp_path}":/data $CLI_IMAGE stop --skip:nightly --skip:pull
```
[the remark this answers is not in the document]
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_backup_restore_tests.bats]: agreed, that should be more stable. and now I;ve added separate test for `stop` command so  I will update this.

File: dockerfiles/cli/tests/cmd_config_tests.bats (hunk 6 above)
Comment (mlabuda) [reply dockerfiles/cli/tests/cmd_config_tests.bats:35]: Sorry, placed on different line, bcs there are more such lines, see in all checks, there are few -e checks. There are -d tests for directories and -f tests for regular files, but there are also -e tests which match to any kind of file or link....

Comment (mlabuda) [reply dockerfiles/cli/tests/cmd_config_tests.bats:35]: Yeah, it should be fine, I was just wondering why it is mixed, whether some of those files are just symlinks or whatever :)

File: dockerfiles/cli/tests/cmd_backup_restore_tests.bats (hunk 7 above)
[the remark this answers is not in the document]
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_backup_restore_tests.bats]: done.

File: dockerfiles/cli/tests/cmd_backup_restore_tests.bats (hunk 4 above)
Comment (riuvshin) [reply dockerfiles/cli/tests/cmd_backup_restore_tests.bats]: done.

**Comment by riuvshin** [comment]
@mlabuda I also working on moving docker run to execute cli command to separate method

**Comment by riuvshin** [comment]
@mlabuda it's done :) now it is way simpler and obvious.

update... 
oops I can't push my commit :( lost write access

**Comment by riuvshin** [comment]
@mlabuda finally pushed :) now it is ready :) 
I've added `execute_cli_command` method to simplify cli commands invocation.

**Review by mlabuda — APPROVED** [review]

**Comment by mlabuda** [comment]
LGTM. Just a question - I have noticed almost every execution of Che CLI has skip pull and skip nightly. Maybe it could be included by default in execution of Che CLI so we avoid more repetition. E.g having DEFAULT_EXTRA_OPTIONS appended to extra options and if user does not provide --che-cli-default-extra-opts then "-skip:nightly --skip:pull" will be used as default.

**Comment by riuvshin** [comment]
@mlabuda yeah... I will take a look on this later. I just want to merge this and then will introduce another PR with more tests and improvements.

**Review by benoitf — APPROVED** [review]
tested and worked on my env

**Comment by riuvshin** [comment]
Thanks guys!

**Pull request merged**